server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/AbstractEndpoint.java

```diff
@@ -0,0 +1,194 @@
+   * the file that it summarizes.
+   */
+  public void invariantViolation(final Response the_response, final String the_log_message) {
+    the_response.status(HttpStatus.BAD_REQUEST_400);
+    Main.LOGGER.error("invariant violation 400: " + the_log_message);
+    my_endpoint_result = the_log_message;
+  }
+
+  /**
+   * Indicate the client cannot perform the requested action because it violates
+   * the server's state machine.
+   */
+  public void illegalTransition(final Response the_response, final String the_log_message) {
+    the_response.status(HttpStatus.FORBIDDEN_403);
+    Main.LOGGER.error("illegal transition attempt 403: " + the_log_message);
+    my_endpoint_result = the_log_message;
+  }
+
+  /**
+   * Indicate that the server is temporarily unavailable. This is typically due
+   * to a long-lived transaction running or server maintenance.
+   */
+  public void serverUnavailable(final Response the_response, final String the_log_message) {
+    the_response.status(HttpStatus.SERVICE_UNAVAILABLE_503);
+    Main.LOGGER.error("server temporarily unavailable 503: " + the_log_message);
+    my_endpoint_result = the_log_message;
+  }
+
+  /**
+   * Indicate that the operation completed successfully.
+   */
+  public void ok(final Response the_response) {
+    the_response.status(HttpStatus.OK_200);    
+  }
+  
+  /**
+   * Indicate and log that the operation completed successfully.
+   */
+  public void ok(final Response the_response, final String the_log_message) {
+    the_response.status(HttpStatus.OK_200);
+    Main.LOGGER.error("successful operation 200: " + the_log_message);
+    my_endpoint_result = the_log_message;
+  }
+
+  /**
+   * This before-filter is evaluated before each request, and can read the
+   * request and read/modify the response. Our before-filter performs
+   * authentication checking.
+   */
+  @Override
+  public void before(final Request the_request, final Response the_response) {
+    // Presume everything goes ok.
+    ok(the_response);
+
+    // Check to see if the server is operating properly.
+    if (!Persistence.hasDB()) {
+      serverError(the_response, "no database");
+      return;
+    }
+    if (Persistence.isTransactionRunning()) {
+      serverUnavailable(the_response, "a long transaction is running");
+      return;
+    }
+    // Check to see if the requested endpoint is permitted from the current
+    // state of the server.
+    if (false) {
+      illegalTransition(the_response, "endpoint not permitted by ASM");
+      return;
+    }
+
+    // This access is well-formed and permitted by the state machine, so log the
+    // use of the endpoint.
+    Main.LOGGER.info("checking authentication on " + the_request);
+
+    // If the client is unauthorized, then indicate such and redirect.
+
+    // If there is no authentication cookie in the request at all, then simply
+    // redirect to the top-level authentication endpoint.
+    // @todo kiniry Add once we have cookies turned on.
```

Comment:
This functionality already works. But doing a redirect to the top-level authentication endpoint doesn't make a whole lot of sense in this context, to me, because the client has to display the authentication page itself - we're not displaying one.